datafusion/core/src/physical_optimizer/sort_enforcement2.rs

```diff
@@ -0,0 +1,2872 @@
+        }
+    } else {
+        false
+    }
+}
+
+fn reverse_window_sort_requirements(
+    request_child: Option<&[PhysicalSortRequirements]>,
+) -> Option<Vec<PhysicalSortRequirements>> {
+    request_child.map(|request| {
+        request
+            .iter()
+            .map(|req| match req.sort_options {
+                None => req.clone(),
+                Some(ops) => PhysicalSortRequirements {
+                    expr: req.expr.clone(),
+                    sort_options: Some(reverse_sort_options(ops)),
+                },
+            })
+            .collect::<Vec<_>>()
+    })
+}
+
+/// Whether to reverse the top WindowExec's sort requirements.
+/// Considering the requirements of the descendants WindowExecs and leaf nodes' output ordering.
+/// TODO！considering all the cases
+fn should_reverse_window_sort_requirements(
+    window_plan: Arc<dyn ExecutionPlan>,
+    top_requirement: Option<&[PhysicalSortRequirements]>,
+    top_reversed_requirement: Option<&[PhysicalSortRequirements]>,
+) -> bool {
+    if top_requirement.is_none() {
+        return false;
+    }
+    let WindowExecInfo { window_expr, .. } =
+        extract_window_info_from_plan(&window_plan).unwrap();
+    let reverse_window_expr = window_expr
+        .iter()
+        .map(|e| e.get_reverse_expr())
+        .collect::<Option<Vec<_>>>();
+    if reverse_window_expr.is_none() {
+        return false;
+    }
+    let flags = window_plan
+        .children()
+        .into_iter()
+        .map(|child| {
+            // If the child is leaf node, check the output ordering
+            if child.children().is_empty()
+                && ordering_satisfy_requirement(
+                    child.output_ordering(),
+                    top_requirement,
+                    || child.equivalence_properties(),
+                )
+            {
+                false
+            } else if child.children().is_empty()
+                && ordering_satisfy_requirement(
+                    child.output_ordering(),
+                    top_reversed_requirement,
+                    || child.equivalence_properties(),
+                )
+            {
+                true
+            } else if child.as_any().downcast_ref::<WindowAggExec>().is_some()
+                || child
+                    .as_any()
+                    .downcast_ref::<BoundedWindowAggExec>()
+                    .is_some()
+            {
+                // If the child is WindowExec, check the child requirements
+                if requirements_compatible(
+                    top_requirement,
+                    child.required_input_ordering()[0].as_deref(),
+                    || child.equivalence_properties(),
+                ) || requirements_compatible(
+                    child.required_input_ordering()[0].as_deref(),
+                    top_requirement,
+                    || child.equivalence_properties(),
+                ) || requirements_compatible(
```

Comment:
This pattern is used a lot. I think you can construct a util to check whether one of the requirements satisfy other (where function does this check). That would be better I think.